aggregation keys doc (#7116)
GoogleChrome/developer.chrome.com: site/en/docs/privacy-sandbox/attribution-reporting/aggregation-keys/index.md

---
layout: 'layouts/doc-post.njk'
title: 'Create aggregation keys'
subhead: >
  What aggregation keys are, how they're used, and how you can translate goals into keys.
description: >
  What aggregation keys are, how they're used, and how you can translate goals into keys.
date: 2022-12-15
updated: 2023-08-25
authors:
  - maudn
---

As an ad tech company running campaigns in multiple locations for various product categories, you want to help advertisers answer the following questions:

1.  How many **purchases** of each product category did each of my campaigns in each geographic region generate?
1.  How much **revenue** for each product category did each of my campaigns in each geographic region generate?

> While many ad tech companies encourage advertisers to configure a variety of conversion types, focusing on the most important conversions such as purchases is a good way to ensure that summary results are detailed and accurate for these important events.

Reviewer: Which important event? I'm a little confused with the logic of this sentence.

To do so, you'll need to think of what questions you want to answer before data is collected.

## Dimensions, keys, and values

To answer these questions, let's take a look at dimensions, keys, and values.

### Dimensions

To understand how your campaigns are generating revenue, as described here, you'll want to track the following dimensions:

> - Ad campaign ID

Reviewer: From a purely style guide perspective, I'd expect all of these to be uniform, ie, 
Dimension name: definition
or
Just dimension name.

Not a hybrid version with both.

- Geography ID: the geographic region where the ad was served
- Product category

While the Campaign ID and the Geography ID dimensions are known when the ad is served (ad-serving time), the Product category will be known from a trigger event, when the user completes a conversion (conversion time).

{% Aside %}
For the diagram in this document, we'll display dimensions as follows:
- Dimensions known at ad-serving time on a green background.
- Dimensions known at conversion time on a yellow background and underlined.
{% endAside %}

> The dimensions you want to track for this example look like this:

Reviewer: ```suggestion
The dimensions you want to track for this example are as shown in the following image:
```

{% Img src="image/RtQlPaM9wdhEJGVKR8boMPkWf443/7H0FNCjZM9zr3VE48Q2G.png", alt="Campaign ID, Geography ID, and product category.", width="498", height="55" %}

### What are aggregation keys (buckets)?

> Aggregation key and bucket refer to the same thing. The term _aggregation key_ is used in the browser APIs used to configure reports. The term _bucket_ is used in the aggregatable and summary reports, and in the [aggregation service APIs](/docs/privacy-sandbox/aggregation-service/).

Reviewer: ```suggestion
The terms aggregation key and bucket refer to the same thing.  _Aggregation key_ is used in the browser APIs used to configure reports. The term _bucket_ is used in the aggregatable and summary reports, and in the [aggregation service APIs](/docs/privacy-sandbox/aggregation-service/).
```

An aggregation key (key for short) is a piece of data that represents the values of the dimensions being tracked. Data is later aggregated along each aggregation key.

For example, let's assume you're tracking the dimensions Product category, Geography ID, and Campaign ID.

> When a user located in Geography ID 7 sees an ad for Campaign ID 12, and later converts by purchasing a product in Product category 25, you may set an aggregation key that looks like this:

Reviewer: ```suggestion
When a user located in Geography ID 7 sees an ad for Campaign ID 12, and later converts by purchasing a product in Product category 25, you may set an aggregation key like the one in the following image:
```

{% Img src="image/RtQlPaM9wdhEJGVKR8boMPkWf443/dmpyjw7KIcunQemDqs8i.png", alt="Aggregation key for a conversion.", width="495", height="125" %}

You'll see later that an aggregation key does not look exactly like this in practice, but for now let's focus on the information contained in the key.

### What are aggregatable values?

> To answer your questions with the dimensions we've outlined, you want to know, across many users and many keys:

Reviewer: I'm confused by this sentence.

Author: fixed

-   The number of purchases (the purchase count). Once aggregated and made available in a summary report, this will be the total purchase count (summary value).
-   The revenue for each purchase (the purchase value). Once aggregated and made available in a summary report, this will be the total revenue (summary value).

Each of these—the purchase count for one conversion and the purchase value for one conversion—is an aggregatable value. You can think of aggregatable values as the values of your measurement goals.

<table class="with-heading-tint">
	<thead>
		<tr><th rowspan="2">Question</th>
			<th rowspan="2">Aggregatable value = Measurement goal
			</th>
        </tr>
	</thead>
	<tbody>
		<tr>
            <td>How many <strong>purchases</strong>…</td>
			<td><strong>Purchase count</strong></td>
		</tr>
		<tr>
			<td>How much <strong>revenue</strong>…</td>
			<td><strong>Purchase value</strong></td>
		</tr>
	</tbody>
</table>

> When a user located in Geography ID 7 sees an ad for Campaign ID 12, and later converts by purchasing a product of Product category 25 for $120, you may set an aggregation key and aggregatable values that look like these:

Reviewer: I actually don't know what we do with currency symbols, but is this written assuming only US ad tech?

-   Aggregatable reports are decrypted by the aggregation service.
-   For each key, the aggregatable values from the aggregatable reports are summed.
-   Noise is added to the summary value.

![image](insert_image_url_here)

The result is a summary report that contains a set of {aggregation key, summary value} pairs.

A summary report contains a JSON dictionary-style set of key-value pairs. Each pair contains:

-   *bucket*: the aggregation key, encoded as a bytestring.
-   *value*: the summary value in decimal for a given measurement goal, summed up from all available aggregatable reports, with an added level of noise.

Example:

ADD CODE BACK IN

In practice, summary reports are encoded in a way that will make buckets and values look different than stated above (that is, a bucket may look like `\u0000\u0000\x80\u0000`). <em>Bucket</em> and <em>value</em> are both bytestrings. 

### Aggregation keys in practice

Aggregation keys are defined by an adtech company, typically in two steps: when an ad is clicked or viewed, and when a user converts.

> {% Aside }

Reviewer 2: This is your problem:

```suggestion
{% Aside %}
```

Author: thanks Barry. completely missed it!
